# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: gfx/gl/SharedSurfaceIO.h
# Commit: 204b899f436d
# Full Hash: 204b899f436d99f0eae759af3ccfcbaabfbbb428
# Author: Jeff Gilbert <jgilbert@mozilla.com>
# Date: 2020-06-11 09:34:54
# Regressor Bug: 1632249
# File Overlap Count: 3
# Description:
#   Bug 1632249 - Support compositing out-of-process WebGL. r=handyman,lsalzman,nical,geckoview-reviewers,agi,imanol
#   
#   * Majorly simplity CanvasRenderer
#   * Replace GLScreenBuffer with trivial GLSwapChain
#   * Use descriptor structs so that future SharedSurface changes aren't so painful
# ==============================================================================

diff -r 3f5baab34b1f -r 204b899f436d gfx/gl/SharedSurfaceIO.h
--- a/gfx/gl/SharedSurfaceIO.h	Thu Jun 11 06:36:55 2020 +0000
+++ b/gfx/gl/SharedSurfaceIO.h	Thu Jun 11 06:37:35 2020 +0000
@@ -14,18 +14,18 @@
 namespace mozilla {
 namespace gl {
 
-class SharedSurface_IOSurface : public SharedSurface {
- private:
+class Texture;
+
+class SharedSurface_IOSurface final : public SharedSurface {
+ public:
+  const UniquePtr<Texture> mTex;
   const RefPtr<MacIOSurface> mIOSurf;
-  GLuint mProdTex;
 
- public:
-  static UniquePtr<SharedSurface_IOSurface> Create(
-      const RefPtr<MacIOSurface>& ioSurf, GLContext* gl, bool hasAlpha);
+  static UniquePtr<SharedSurface_IOSurface> Create(const SharedSurfaceDesc&);
 
  private:
-  SharedSurface_IOSurface(const RefPtr<MacIOSurface>& ioSurf, GLContext* gl,
-                          const gfx::IntSize& size, bool hasAlpha);
+  SharedSurface_IOSurface(const SharedSurfaceDesc&, UniquePtr<MozFramebuffer>,
+                          UniquePtr<Texture>, const RefPtr<MacIOSurface>&);
 
  public:
   ~SharedSurface_IOSurface();
@@ -36,29 +36,9 @@
   virtual void ProducerAcquireImpl() override {}
   virtual void ProducerReleaseImpl() override;
 
-  virtual bool CopyTexImage2D(GLenum target, GLint level, GLenum internalformat,
-                              GLint x, GLint y, GLsizei width, GLsizei height,
-                              GLint border) override;
-  virtual bool ReadPixels(GLint x, GLint y, GLsizei width, GLsizei height,
-                          GLenum format, GLenum type, GLvoid* pixels) override;
-
-  virtual GLuint ProdTexture() override { return mProdTex; }
-
-  virtual GLenum ProdTextureTarget() const override {
-    return LOCAL_GL_TEXTURE_RECTANGLE_ARB;
-  }
-
-  static SharedSurface_IOSurface* Cast(SharedSurface* surf) {
-    MOZ_ASSERT(surf->mType == SharedSurfaceType::IOSurface);
-    return static_cast<SharedSurface_IOSurface*>(surf);
-  }
-
-  MacIOSurface* GetIOSurface() const { return mIOSurf; }
-
   virtual bool NeedsIndirectReads() const override { return true; }
 
-  virtual bool ToSurfaceDescriptor(
-      layers::SurfaceDescriptor* const out_descriptor) override;
+  Maybe<layers::SurfaceDescriptor> ToSurfaceDescriptor() override;
 
   virtual bool ReadbackBySharedHandle(
       gfx::DataSourceSurface* out_surface) override;
@@ -66,25 +46,18 @@
 
 class SurfaceFactory_IOSurface : public SurfaceFactory {
  public:
-  // Infallible.
-  static UniquePtr<SurfaceFactory_IOSurface> Create(
-      GLContext* gl, const SurfaceCaps& caps,
-      const RefPtr<layers::LayersIPCChannel>& allocator,
-      const layers::TextureFlags& flags);
-
- protected:
   const gfx::IntSize mMaxDims;
 
-  SurfaceFactory_IOSurface(GLContext* gl, const SurfaceCaps& caps,
-                           const RefPtr<layers::LayersIPCChannel>& allocator,
-                           const layers::TextureFlags& flags,
-                           const gfx::IntSize& maxDims)
-      : SurfaceFactory(SharedSurfaceType::IOSurface, gl, caps, allocator,
-                       flags),
-        mMaxDims(maxDims) {}
+  explicit SurfaceFactory_IOSurface(GLContext& gl);
 
-  virtual UniquePtr<SharedSurface> CreateShared(
-      const gfx::IntSize& size) override;
+  virtual UniquePtr<SharedSurface> CreateSharedImpl(
+      const SharedSurfaceDesc& desc) override {
+    if (desc.size.width > mMaxDims.width ||
+        desc.size.height > mMaxDims.height) {
+      return nullptr;
+    }
+    return SharedSurface_IOSurface::Create(desc);
+  }
 };
 
 }  // namespace gl